### DISCUSSION TITLE
Workaround for User Functions in Mobile

### DISCUSSION THREAD

**Discussion by scwunch** [SilentVoid13/Templater#453] [Templates Showcase]
I just discovered something that made me happy and I would like to share it.  

I was disappointed when I found out that currently, [user functions are unavailable on mobile](https://silentvoid13.github.io/Templater/docs/user-functions).  Then I remembered the provided [tp.file.include](https://silentvoid13.github.io/Templater/docs/internal-variables-functions/internal-modules/file-module) function.  I thought maybe I could use it to import a function definition to use in the current template.  Alas, it appears to only import the processed output and discards the code.  So there is no way to pass arguments to those functions... *or is there?*

It turns out I can pass arguments as properties of any predefined object (like `tp`) and those properties will be accessible by any template called within the "tp.file.include" statement.  Therefore, we can effectively pass arguments to functions like so: 

### Example: 
Templates/Greet me by name.md
```
<%*
tp.name = "Ryan";
let greeting = tp.file.include([[Templates/Say Hello.md]]);
tR += greeting;
%>
```

Templates/Say Hello.md
```
<%*
return "Hello "+tp.name;
%>
```
or equivalently: 
```
Hello <% tp.name %>
```

Now wherever "Greet me by name.md" is used, it will produce "Hello Ryan".  **And it works in mobile!**

Any valid name can be appended to `tp.` and any value can be assigned to it — just be careful not to overwrite existing members of `tp` or you may get unexpected results!  Alternatively, you could resolve to always pass arguments as arrays to `tp.args` or something similar.

**Comment by scwunch** [comment]
I just found an easier way of doing this, much less hacky 😅

In the template file containing the function you want to export, include the line `return myFunction;`  When you try to resolve this file on it's own, it will give you an error.  However, if you `include` it in another template via `const myFunction = await tp.file.include('[[My Function]]')` for use in another template it will work as expected.

**Reply by paulrudy** [reply]
Thank you so much for this, @scwunch !

## Edit: incorrect info below. Instead see [my later comment](https://github.com/SilentVoid13/Templater/discussions/453#discussioncomment-2030579) or @cfritze64 's [comment](https://github.com/SilentVoid13/Templater/discussions/453#discussioncomment-2028086) for a different working example.

For anyone who wasn't exactly clear on how to implement this:

Template file `Templater Send Function.md`:
```
<%* function myFunction() {
let foo = "bar";
}
return myFunction;
%>
```
*or for the same result but without defining a full function when one isn't needed*, Template file `Templater Send Function.md`:
```
<%*
let foo = "bar";
return foo;
%>
```

Template file `Templater Receive Function.md`:
```
<%* 
await tp.file.include('[[Templater Send Function]]');
tR += "foo is: " + foo;
%>

once again, foo is: <% foo %>
```
(note `const myFunction = await`… does not seem necessary. Just `await`…

Then creating a file from `Templater Receive Function.md` results in file `Untitled.md`:
```
foo is: bar

once again, foo is: bar
```

**Reply by eb4i** [reply]
Hello Everyone, and thanks for sharing this solution. 

I've been unsuccessful in implementing @paulrudy approach and I'm unable to determine why. Templater only creates a new empty file when the wrapper template is called, see the below image. I've confirmed that Templater is working correctly and have tried many combinations; the issue seems related to the wrapper template using a JavaScript execution command with await. Please can you tell me _how to investigate the root cause of this issue_, are there any _viewable logs with iOS Obsidian_ or how to troubleshoot the issue? 

Thank you, and very much appreciate any assistance. 

UPDATE: 
Using the Desktop version, the console logs this error:
`Log.ts:17 Templater Error: Template parsing error, aborting. 
 foo is not defined`

Config
![797C30EC-BC3E-48FE-AECE-006DFD71562C](https://user-images.githubusercontent.com/3425229/150661209-8ac1fe90-1011-4862-b0eb-c1fa9cb3f5dd.png)

Obsidian version: 1.0.5
iOS: 15.2.1

**Reply by cfritze64** [reply]
i had to adjust the approach a little. here are my notes from when i got it working:

the file addOne.md contains my user function. it is a regular Obsidian markdown file in my normal folder structure:

<%* 
let added = [];
for (const num of tp.args) {
added.push(num+1);
}
return added;
%>


and the template file for Templater resides in the designated templates folder for the plugin. it contains:

<%*
tp.args = [1,2,3];
const myFunc = await tp.file.include('[[addOne]]');
%>

<% myFunc %>

note that other functions in the templater tp object are also accessible, like tp.system.prompt()

**Reply by eb4i** [reply]
@cfritze64 impressive 👏

**Reply by shabegom** [reply]
1.10.0 should now natively support JS scripts without needing any workarounds.

**Comment by paulrudy** [comment]
My initial comment was incorrect. See below for an updated version that actually works. Or see @cfritze64 's [comment](https://github.com/SilentVoid13/Templater/discussions/453#discussioncomment-2028086) for another working example.

The difference between what I initially posted and the working code below is that the sender function needs to store `foo` within the `tp` object as `tp.foo`, and it doesn't need to return anything. The receiver function needs then to also refer to `tp.foo`:


Template file `Templater Send Function.md`:
```
<%* function myFunction() {
tp.foo = "bar";
}
%>
```
*or for the same result but without defining a full function when one isn't needed*, Template file `Templater Send Function.md`:
```
<%*
tp.foo = "bar";
%>
```

Template file `Templater Receive Function.md`:
```
<%* 
await tp.file.include('[[Templater Send Function]]');
tR += "foo is: " + tp.foo;
%>

once again, foo is: <% tp.foo %>
```
(note `const myFunction = await`… does not seem necessary. Just `await`…

Then creating a file from `Templater Receive Function.md` results in file `Untitled.md`:
```
foo is: bar

once again, foo is: bar
```

@eb4i to troubleshoot in macOS, use cmd-opt-i to view the console. I don't know how to do it on other platforms, unfortunately

**Reply by eb4i** [reply]
@paulrudy works like a dream, and thank you for replying.
